sdk/python/kfp/components/structures/components.json_schema.json

```diff
@@ -0,0 +1,402 @@
+{
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "$id": "http://kubeflow.org/pipelines/components.json_schema.json",
+
+  "$ref": "#/definitions/PipelineRunSpec",
+
+  "definitions": {
+    "PrimitiveTypes": {
+      "oneOf": [
+        {"type": "string"},
+        {"type": "integer"},
+        {"type": "number"},
+        {"type": "boolean"}
+      ]
+    },
+
+    "TypeType": {
+      "oneOf": [
+        {"type": "string"},
+        {"type": "array", "items": {"$ref": "#/definitions/TypeType"}},
+        {"type": "object", "additionalProperties": {"$ref": "#/definitions/TypeType"}}
+      ]
+    },
+
+    "InputSpec": {
+      "description": "Describes the component input specification",
+      "type": "object",
+      "properties": {
+        "name": {"type": "string"},
+        "type": {"$ref": "#/definitions/TypeType"},
+        "description": {"type": "string"},
+        "default": {"$ref": "#/definitions/PrimitiveTypes"},
+        "optional": {"type": "boolean", "default": false}
+      },
+      "additionalProperties": false
+    },
+
+    "OutputSpec": {
+      "description": "Describes the component output specification",
+      "type": "object",
+      "required": ["name"],
+      "properties": {
+        "name": {"type": "string"},
+        "type": {"$ref": "#/definitions/TypeType"},
+        "description": {"type": "string"}
+      },
+      "additionalProperties": false
+    },
+
+    "InputValuePlaceholder": {
+      "description": "Represents the command-line argument placeholder that will be replaced at run-time by the input argument value.",
+      "type": "object",
+      "required": ["inputValue"],
+      "properties": {
+        "inputValue" : {
+          "description": "Name of the input.",
+          "type": "string"
+        }
+      },
+      "additionalProperties": false
+    },
+
+    "InputPathPlaceholder": {
+      "description": "Represents the command-line argument placeholder that will be replaced at run-time by a local file path pointing to a file containing the input argument value.",
+      "type": "object",
+      "required": ["inputPath"],
+      "properties": {
+        "inputPath" : {
+          "description": "Name of the input.",
+          "type": "string"
+        }
+      },
+      "additionalProperties": false
+    },
+
+    "OutputPathPlaceholder": {
+      "description": "Represents the command-line argument placeholder that will be replaced at run-time by a local file path pointing to a file where the program should write its output data.",
```

Comment:
If this is a command line spec placeholder, do we need to draw a distinction between input and output?

Reply:
Interesting point.
The difference is that `inputPath` references and input while `outputPath` references an output. Output can have same names as input, so there can be conflicts if we do not distinguish outputs from inputs.